### PR TITLE
Implemented the redesign of the inventory database.

### PR DESCRIPTION
**Pull request by Talos10**
# Description

Implemented the redesign of the inventory database.

Issue reference: #21 

## Type of change

Please delete options that are not relevant.

- [ ] BUG (Fixing or documenting a bug)
- [ ] FEAT (Add a new feature to the project)
- [ ] CHORE (Something that needs to be done outside of the issue)
- [ ] DOCS (Implement documentation)
- [X ] TASK (A task that must be done to complete an issue)

# Checklist:

- [X] My pull request references the issue in the comments
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: scripts/createDB.sql [hunk 1]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
+    REFERENCES `soen_390_db`.`inventory_item` (`inventory_item_id`)
+    ON DELETE NO ACTION
+    ON UPDATE NO ACTION,
+  CONSTRAINT `notMadeFromSameItem`
+    CHECK (`inventory_item_id` != `made_from_inventory_item_id`))
+    
+	collate = utf8mb4_unicode_ci;
+
+-- Table to store the raw materials needed to build the semi-finished goods.
 CREATE TABLE `soen_390_db`.`raw_material` (
```
Comment (phong1233) [inline scripts/createDB.sql]: We should change the name raw material to raw good that way we have finished_good, semi_finished_good and raw_good

File: scripts/createDB.sql [hunk 2]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
   
   collate = utf8mb4_unicode_ci;
-    
+
+-- Table to store the semi-finished goods needed to build the finished goods.
 CREATE TABLE `soen_390_db`.`semi_finished_good` (
   `semi_finished_good_id` INT NOT NULL AUTO_INCREMENT,
-  `semi_finished_good_name` VARCHAR(45) NOT NULL,
-  `quantity` INT NOT NULL,
+  `inventory_item_id` INT NOT NULL,
```
Comment (phong1233) [inline scripts/createDB.sql]: is it possible to only keep inventory_item_id and remove semi_finished_good id?

File: scripts/createDB.sql [hunk 3]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
+    REFERENCES `soen_390_db`.`inventory_item` (`inventory_item_id`)
+    ON DELETE CASCADE
+    ON UPDATE NO ACTION)
   
   collate = utf8mb4_unicode_ci;
-    
+
+-- Table to store the finished goods that will be sold to the customers.
 CREATE TABLE `soen_390_db`.`finished_good` (
   `finished_good_id` INT NOT NULL AUTO_INCREMENT,
```
Comment (phong1233) [inline scripts/createDB.sql]: same here only keep inventory_item_id if possible

File: scripts/createDB.sql [hunk 4]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
-    REFERENCES `soen_390_db`.`raw_material` (`raw_material_id`)
-    ON DELETE NO ACTION
-    ON UPDATE CASCADE)
-    
-	collate = utf8mb4_unicode_ci;
 
-CREATE TABLE `soen_390_db`.`property_of_raw_material` (
-  `raw_material_id` INT NOT NULL,
+-- Table to store the various properties of the inventory items.
+CREATE TABLE `soen_390_db`.`property_of_inventory_item` (
```
Comment (phong1233) [inline scripts/createDB.sql]: I think property_of_good is a better name for the table

File: scripts/createDB.sql [hunk 5]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
-CREATE TABLE `soen_390_db`.`property_of_semi_finished_good` (
-  `semi_finished_good_id` INT NOT NULL,
-  `property_name` VARCHAR(45) NOT NULL,
-  `property_value` VARCHAR(45) NOT NULL,
-  PRIMARY KEY (`semi_finished_good_id`, `property_name`),
-  CONSTRAINT `semiFinishedGoodPropertyIDForeignKey`
-    FOREIGN KEY (`semi_finished_good_id`)
-    REFERENCES `soen_390_db`.`semi_finished_good` (`semi_finished_good_id`)
+-- Table to store the various raw materials that make up each raw material order.
+CREATE TABLE `soen_390_db`.`ordered_raw_material` (
```
Comment (phong1233) [inline scripts/createDB.sql]: This should be removed and we should keep only 1 table for ordered_good

File: scripts/createDB.sql [hunk 6]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
-    FOREIGN KEY (`finished_good_id`)
-    REFERENCES `soen_390_db`.`finished_good` (`finished_good_id`)
-    ON DELETE CASCADE
-    ON UPDATE NO ACTION)
-    
-	collate = utf8mb4_unicode_ci;
-    
+-- Table to store the various semi-finished goods that production has given an order to
+-- produce for each production order.
 CREATE TABLE `soen_390_db`.`ordered_semi_finished` (
```
Comment (phong1233) [inline scripts/createDB.sql]: same here, only keep 1 table for all item orders

File: scripts/createDB.sql [hunk 7]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
+    FOREIGN KEY (`inventory_item_id`)
+    REFERENCES `soen_390_db`.`inventory_item` (`inventory_item_id`)
     ON DELETE NO ACTION
     ON UPDATE CASCADE)
     
 	collate = utf8mb4_unicode_ci;
 
+-- Table to store the various finished goods that a customer has placed an order for
+-- for each customer order.
 CREATE TABLE `soen_390_db`.`ordered_finished_by_customer` (
```
Comment (phong1233) [inline scripts/createDB.sql]: This table isn't necessary for now

File: scripts/createDB.sql [hunk 8]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
+    REFERENCES `soen_390_db`.`inventory_item` (`inventory_item_id`)
     ON DELETE NO ACTION
     ON UPDATE CASCADE)
     
 	collate = utf8mb4_unicode_ci;
-    
+
+-- Table to store the various finished goods that production has given an order to
+-- produce for each production order.
 CREATE TABLE `soen_390_db`.`ordered_finished_by_production` (
```
Comment (phong1233) [inline scripts/createDB.sql]: We should keep only 1 table for ordered items

File: scripts/createDB.sql [hunk 9]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
   `date_time_added` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
   `date_time_finished` DATETIME,
-  `price` DECIMAL(10,2) NOT NULL,
   PRIMARY KEY (`production_order_id`))
   
   collate = utf8mb4_unicode_ci;
 
+-- Table to store the raw materials, semi-finished goods, and finished goods all
+-- under one table.
+CREATE TABLE `soen_390_db`.`inventory_item` (
```
Comment (phong1233) [inline scripts/createDB.sql]: We should change the name to inventory_good to match the other tables

File: scripts/createDB.sql [hunk 10]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
...
+  `inventory_item_type` VARCHAR(45) NOT NULL,
+  `quantity` INT NOT NULL DEFAULT 0,
+  `date_time_added` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  PRIMARY KEY (`inventory_item_id`),
+  CONSTRAINT `validItemType`
+    CHECK (`inventory_item_type` IN ("raw", "semi-finished", "finished")))
+  
+  collate = utf8mb4_unicode_ci;
+
+CREATE TABLE `soen_390_db`.`inventory_item_composed_of` (
```
Comment (phong1233) [inline scripts/createDB.sql]: Could name this table composition_of_good

File: scripts/createDB.sql [hunk 11]
```diff
@@ -21,7 +22,8 @@ CREATE TABLE `soen_390_db`.`user` (
   PRIMARY KEY (`userID`))
 
   collate = utf8mb4_unicode_ci;
-  
+
+-- Table to store the customer orders (orders of bikes).
 CREATE TABLE `soen_390_db`.`customer_order` (
```
Comment (phong1233) [inline scripts/createDB.sql]: I don't think this table is necessary for now

File: scripts/createDB.sql [hunk 12]
```diff
@@ -31,302 +33,326 @@ CREATE TABLE `soen_390_db`.`customer_order` (
   PRIMARY KEY (`customer_order_id`))
   
   collate = utf8mb4_unicode_ci;
+
+-- Table to store the different orders of raw materials needed for producing bikes.
+CREATE TABLE `soen_390_db`.`raw_material_order` (
```
Comment (phong1233) [inline scripts/createDB.sql]: not necessary we should keep only 1 table for all goods' order

**Review by phong1233 — APPROVED** [review]
Very nice!

**Review by sebastien-blain — APPROVED** [review]
Good job, db looks nice

**Pull request merged**